test/AshToken.js

```diff
@@ -0,0 +1,174 @@
+const chai = require("chai");
+const chaiAsPromised = require("chai-as-promised");
+const { expect } = chai;
+chai.use(chaiAsPromised);
+const BN = require("bn.js");
+chai.use(require("chai-bn")(BN));
+const { expectEvent } = require("openzeppelin-test-helpers");
+
+const ashToken = artifacts.require("./AshToken.sol");
+
+const { makeRandomAddress } = require("./utils");
+
+contract("AshToken", accounts => {
+  let testContract;
+
+  before(async () => {
+    testContract = await ashToken.deployed();
+  });
+
+  describe("decimals()", () => {
+    it("has 12 decimals", async () => {
+      const result = await testContract.decimals();
+      expect(result).to.be.a.bignumber.that.equals("12");
+    });
+  });
+
+  describe("symbol()", () => {
+    it("has symbol ASH", async () => {
+      const result = await testContract.symbol();
+      expect(result).to.eql("ASH");
+    });
+  });
+
+  describe("name()", () => {
+    it("has name Ash Token", async () => {
+      const result = await testContract.name();
+      expect(result).to.eql("Ash Token");
+    });
+  });
+
+  describe("totalSupply()", () => {
+    it("has initial supply of 0", async () => {
+      const result = await testContract.totalSupply();
+      expect(result).to.be.a.bignumber.that.equals("0");
+    });
+  });
+
+  describe("balanceOf()", () => {
+    it("has initial balance of 0", async () => {
+      const address = makeRandomAddress();
+      const balance = await testContract.balanceOf(address);
+      expect(balance).to.be.a.bignumber.that.equals("0");
+    });
+  });
+
+  describe("mint()", () => {
+    it("works", async () => {
+      const address = makeRandomAddress();
+      await testContract.mint(address, 7);
+
+      const balance = await testContract.balanceOf(address);
+      expect(balance).to.be.a.bignumber.that.equals("7");
+
+      const totalSupply = await testContract.totalSupply();
+      expect(totalSupply).to.be.a.bignumber.that.equals("7");
+    });
+  });
+
+  describe("transfer()", () => {
+    const owner = accounts[0];
+
+    before(async () => {
+      await testContract.mint(accounts[0], 4);
+    });
+
+    it("works", async () => {
+      const recipient = makeRandomAddress();
+      const { tx } = await testContract.transfer(recipient, 4);
+
+      await expectEvent.inTransaction(tx, ashToken, "Transfer", {
+        // Event argument names not standardized across ERC20 implementation. Use names from contract here.
```

Comment:
Interesting comment. Is there a way to query over multiple contracts with different event arguments?

Reply:
I guess the trick is that in raw Ethereum those are positional arguments, so the names do not matter that much. Inspecting the logs object in truffle leads to

```
    event: 'Transfer',
    args:
     Result {
       '0': '0x627306090abaB3A6e1400e9345bC60c78a8BEf57',
       '1': '0xAe2A56e5014D7DF7b8619D708E8Adc3C6806Cd85',
       '2': [BN],
       __length__: 3,
       from: '0x627306090abaB3A6e1400e9345bC60c78a8BEf57',
       to: '0xAe2A56e5014D7DF7b8619D708E8Adc3C6806Cd85',
       value: [BN] }
```

so we might consider checking "0", "1", "2" instead of the names.

Reply:
Nice point.
Great to use names for testing this contract, but I guess we use numbers for erc20 support in iov-core.